core/src/mbc/mbc3.rs

```diff
@@ -0,0 +1,263 @@
+use std::cmp::max;
+
+use crate::{
+    log::*,
+    cartridge::{RamSize, RomSize},
+    primitives::{Byte, Word},
+};
+use super::Mbc;
+
+/// Third version of the memory bank controller. In contrast to all other MBCs,
+/// this one can have a real time clock (RTC).
+pub(crate) struct Mbc3 {
+    rom: Box<[Byte]>,
+    ram: Box<[Byte]>,
+
+    /// Stores the current ROM bank. 7 bits are usable, the MSB is always 0.
+    /// While all bits are writeable, bank 0 still cannot be selected (this is
+    /// checked when accessing the ROM).
+    rom_bank: u8,
+
+    /// This stores the current RAM bank OR the current RTC register. A value
+    /// of 0--3 enables a specific RAM bank; a value from 8--C enables a RTC
+    /// register.
+    ram_bank: u8,
+
+    /// Whether or not the RAM and RTC registers are enabled.
+    ram_enabled: bool,
+
+    /// This serves a dual role.
+    ///
+    /// For reading, the user has to latch the registers. That means that the
+    /// actual values from the clock are written into this value (as some kind
+    /// of cache). The clock continues to run in the background.
+    ///
+    /// For writing, the user has to set the HALT flag to `true`. As soon as
+    /// that flag is set to `false` again, the values are "committed" to the
+    /// actual clock.
+    ///
+    /// TODO: currently, the actual clock is not implemented. The registers
+    /// always return 0.
+    rtc_regs: RtcRegisters,
+
+    /// When the user writes a 0 and then a 1 into this register, the clock's
+    /// values are latched into the RTC registers.
+    latch_rtc: Byte,
+}
+
+
+impl Mbc3 {
+    pub(crate) fn new(data: &[u8], rom_size: RomSize, ram_size: RamSize) -> Self {
+        assert!(rom_size <= RomSize::Banks128, "More than 128 banks, but only MBC3!");
+        assert!(
+            rom_size.len() == data.len(),
```

Comment:
Does this assertion make sense? In my opinion the `data` needs to fit in the `rom_size` but the size does not have to match exactly.

Reply:
I wasn't sure about that either. I decided to add the stricter assert to see what happens. It seems that ROM files always have exactly the correct size, even if part of that data is not used (probably just 0s). We can of course relax this assertion, but I'm kind of curious to see if there are ROMs where this does not match.

Reply:
That's interesting! In this case I like your implementation - let's leave it this way.